### PR TITLE
Lab4 Хо Куанг Чунг

### PR DESCRIPTION
**Pull request by gnurtqh**
Основные и дополнительное задание.

### CODE REVIEW

**Comment by medbar** [comment]
Добрый вечер. В последнем задании, где надо было построить графики, строить график атрибута .time бессмысленно, так как этот атрибут всегда равен длине входной последовательности. Построй графики времени выполнения поиска гипотезы . 

В дополнительном задании не правильно посчитан WER. Между файлами надо усреднять не WER, а надо усреднять insertion/deletion/substitution/correction. В документации jiwer есть пример как считать корректно на большом количестве произнесений.

**Comment by gnurtqh** [comment]
@medbar Я исправил ошибки

**Comment by gnurtqh** [comment]
Для нахождения лучшей гипотезы распознавания можно использовать:
- Greedy Search: Greedy Search аналогичен алгоритм Beam Search с beam_size=1 - этот алгоритм просто выбирает наилучший вариант на каждом шаге
- Вариационный поиск (Viterbi algorithm): Этот алгоритм учитывает все возможные пути в графе гипотез, чтобы найти наилучший. Он более точный, чем гранично-ориентированный поиск, но также и более сложный в реализации

Beam search считается наилучшим выбором для этой задачи, потому что Beam search позволяет нам сбалансировать точность и скорость поиска через параметр beam_size. Мы можем настраивать этот параметр, чтобы он соответствовал конкретным задачам.

**Comment by medbar** [comment]
Витерби точно можно использовать для поиска лучшего результата распознавания? Объясни это подробнее, пожалуйста

**Comment by medbar** [comment]
Все верно, A* может быть хорошей заменой Beam Search. баллы выставил.

**Comment by medbar** [comment]
@gnurtqh, двое других студентов сдали работу раньше тебя, поэтому как и договаривались они получили автомат за экзамен. Поскольку ты создал PR раньше всех (хоть и с ошибкой в последнем задании), я считаю, что ты тоже заслуживаешь автомат. Поэтому даю тебе дополнительное задание. Если выполнишь его, получишь автоматом 5 за экзамен. 

Надо обучить хорошую акустическую модель (asr_lab3) и сгенерировать новый resources/lab4/test_am_nlogprobs.ark, такой, чтобы в финальном задании asr_lab4 WER значительно улучшился. 


Вот код, который генерирует test_am_nlogprobs.ark
```python
from kaldiio import WriteHelper
fe = FeatureExtractor()
test_dataset = FeatsPhoneDataset('timit/data/TEST/', feature_extractor=fe, phone2id=PHONE2ID)
test_dataloader = torch.utils.data.DataLoader(test_dataset, batch_size=1, 
                                               num_workers=0, collate_fn=test_dataset.collate_pad, shuffle=False)

results = torch.load('baseline.trainer')
results.to('cuda')
limit=40
with torch.no_grad(), WriteHelper('ark,t:resources/lab4/test_am_nlogprobs.ark') as f:
    i=0
    for b in tqdm(test_dataloader, total=limit):
        b = results.batch_to_device(b)
        logits = results.model(**b)
        f(b['uri'][0], - torch.log_softmax(logits.squeeze(dim=1), dim=-1).cpu().numpy())
        i+=1
        if i == limit:
            break
    
```

**Comment by gnurtqh** [comment]
Большое спасибо, я сделаю это задание.

**Comment by gnurtqh** [comment]
@medbar Я уменьшил потери модели до 1.11, однако похоже, что баллы WER не улучшились. Я думаю, что для обучения достаточно хорошей модели трансформера требуется больше данных, или же может потребоваться использование предварительно обученных моделей для достижения лучших результатов.

**Comment by gnurtqh** [comment]
@medbar Я попытался улучшить модель lab3 до loss = 1,11, однако WER существенно не улучшился (~ 0,8).
Я попытался заменить TransformerEncoderLayer в AModel на предобученную модель wav2vec2 и сгенерировать новый resources/lab4/test_am_nlogprobs.ark. WER улучшился до ~0,5

**Comment by medbar** [comment]
Спасибо, считаю что результат хороший! Твоя AM существенно лучше бейзлайна. Создайте, пожалуйста, отдельный PR c обновлением resources/lab4/test_am_nlogprobs.ark. Вольем его в master и я отпущу вас с автоматом за экзамен.
